python/cudf/cudf/_lib/copying.pyx

```diff
@@ -735,3 +737,120 @@ def segmented_gather(Column source_column, Column gather_map):
 
     result = Column.from_unique_ptr(move(c_result))
     return result
+
+
+cdef class PackedColumns:
+
+    def __reduce__(self):
+        return self.deserialize, self.serialize()
+
+    @property
+    def __cuda_array_interface__(self):
+        cdef dict intf = {
+            "data": (self.gpu_data_ptr, False),
+            "shape": (self.gpu_data_size,),
+            "strides": None,
+            "typestr": "|u1",
+            "version": 0
+        }
+        return intf
+
+    @property
+    def metadata_ptr(self):
+        return int(<uintptr_t>self.c_metadata_ptr())
+
+    @property
+    def metadata_size(self):
+        return int(self.c_metadata_size())
+
+    @property
+    def gpu_data_ptr(self):
+        return int(<uintptr_t>self.c_gpu_data_ptr())
+
+    @property
+    def gpu_data_size(self):
+        return int(self.c_gpu_data_size())
+
+    def serialize(self):
+        header = {}
+
+        header["column-names"] = self.column_names
+        header["index-names"] = self.index_names
+        header["gpu-data-ptr"] = self.gpu_data_ptr
+        header["gpu-data-size"] = self.gpu_data_size
+        header["metadata"] = list(
+            <uint8_t[:self.metadata_size]>self.c_metadata_ptr()
+        )
```

Comment:
Would it be possible to pass through only the metadata's pointer and size here, and reconstruct the vector during deserialization? I tried out something like

```cython
cdef vector[uint8_t] v = list(<uint8_t[:header["metadata-size"]]><void*>header["metadata-ptr"])
```

To achieve this but it resulted in a vector of garbled data.

Reply:
I am guessing the entire metadata array needs to be copied as the pointer would point somewhere random in the destination memory space. It lives on the host though so that should be doable right?